### ISSUE TITLE
A way to provide fitness with direct confrontation

### ISSUE THREAD

**Issue by meditans** [raymoo/NEET#6]
It would be nice to define a grading function for a genome that is depends on his performance in direct confrontation with others in his subspecies
Let's say based on a function:

```
fight :: MonadRandom m => Genome -> Genome -> m (score, score)
```

How would you accomplish this with the current version of the library?

**Comment by raymoo** [comment]
I don't think it is possible currently, since the TrainMethod passed to the training function has no way to look at what species each genome belongs to. Maybe it would be useful for TrainMethods to receive more metadata, to enable things like this.

Once it is possible to know which species each genome is in, one way might be to collect the members of each species in different bins, and evaluate fight for every pairing, add up (for each genome) all of its scores, and divide it by the number of confrontations.

The last issue would be (something that I unfortunately didn't document yet) that the `t Double` output by the `TrainMethod` needs to be in the same order as the input `t Genome`. One, maybe hackish,  way to do this, would be to create an intermediate `t (Int, Genome)` giving each Genome an index. After finding the scores for each Genome, you could go through each `(Int, Genome)` and map it to the score you found for that index.
